packages/mongodb-chatbot-eval/src/Pipeline.ts

```diff
@@ -0,0 +1,134 @@
+  report: PipelineReportFunc
+) => Promise<void>;
+
+interface RunPipelineParams {
+  /**
+    A function that returns an {@link EvalConfig}.
+    This can be the same function as what you export from your configuration file.
+   */
+  configConstructor: ConfigConstructor;
+
+  /**
+    A function that takes `generate`, `evaluate`, and `report` functions
+    based on your configuration and runs a pipeline of commands.
+
+    These functions have the same behavior as the CLI commands,
+    creating the relevant data in their respective stores
+    and adding the command metadata.
+   */
+  pipelineFunc: PipelineFunc;
+}
+
+/**
+  Runs a pipeline of commands from configuration file.
+  This is a useful utility for chaining a group of commands.
+
+  For example, you may want to generate one set of converations,
+  then generate N evaluations for each conversation,
+  and then M reports for each evaluation.
+ */
+export const runPipeline = async ({
+  configConstructor,
+  pipelineFunc,
+}: RunPipelineParams): Promise<void> => {
+  logger.info("Starting pipeline.");
+  logger.info("Constructing pipeline config.");
+
+  const config = await configConstructor();
+  const {
+    metadataStore,
+    generatedDataStore,
+    evaluationStore,
+    reportStore,
+    commands,
+    afterAll,
+  } = config;
+
+  const generateFunc: PipelineGenerateFunc = async (name) => {
+    assert(
+      commands.generate?.[name],
+      `Cannot find generate command with the name '${name}'.`
+    );
+    const { testCases, generator } = commands.generate[name];
+    const { metadata } = await generateDataAndMetadata({
+      testCases,
+      name,
+      generator,
+      generatedDataStore: generatedDataStore,
+      metadataStore: metadataStore,
+    });
+    return metadata;
+  };
+
+  const evaluateFunc: PipelineEvaluateFunc = async (
+    name,
+    generatedDataRunId
+  ) => {
+    assert(
+      commands.evaluate?.[name],
+      `Cannot find evaluate command with the name '${name}'.`
+    );
+    const { evaluator } = commands.evaluate[name];
+    const { metadata } = await generateEvalsAndMetadata({
+      name,
+      evaluator,
+      generatedDataRunId,
+      generatedDataStore: generatedDataStore,
+      evaluationStore: evaluationStore,
+      metadataStore: metadataStore,
+    });
+    return metadata;
+  };
+
+  const reportFunc: PipelineReportFunc = async (name, evaluationRunId) => {
+    assert(
+      commands.report?.[name],
+      `Cannot find report command with the name '${name}'.`
+    );
+    const { reporter } = commands.report[name];
+    const { metadata } = await generateReportAndMetadata({
+      name,
+      reportEvalFunc: reporter,
+      reportStore: reportStore,
+      evaluationStore: evaluationStore,
+      metadataStore: metadataStore,
+      evaluationRunId,
+    });
+    return metadata;
+  };
+
+  try {
+    logger.info("Running pipeline actions.");
+    await pipelineFunc(generateFunc, evaluateFunc, reportFunc);
+  } finally {
```

Comment:
TODO: the function isn't exiting. the err:

```
│ 2024-03-04T19:52:09.422186959Z eval-quality-manual-j98-xbtcz /app/packages/mongodb-rag-core/node_modules/mongodb/lib/sdam/topology.js:217                                                                    │
│ 2024-03-04T19:52:09.422208186Z eval-quality-manual-j98-xbtcz             drainWaitQueue(this[kWaitQueue], new error_1.MongoTopologyClosedError());                                                           │
│ 2024-03-04T19:52:09.422242836Z eval-quality-manual-j98-xbtcz                                              ^                                                                                                  │
│ 2024-03-04T19:52:09.422248401Z eval-quality-manual-j98-xbtcz MongoTopologyClosedError: Topology is closed                                                                                                    │
│ 2024-03-04T19:52:09.422261465Z eval-quality-manual-j98-xbtcz     at /app/packages/mongodb-rag-core/node_modules/mongodb/lib/sdam/topology.js:217:46                                                          │
│ 2024-03-04T19:52:09.422268125Z eval-quality-manual-j98-xbtcz     at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {                                                             │
│ 2024-03-04T19:52:09.422275116Z eval-quality-manual-j98-xbtcz   [Symbol(errorLabels)]: Set(0) {}                                                                                                              │
│ 2024-03-04T19:52:09.422281944Z eval-quality-manual-j98-xbtcz }                                                                                                                                               │
│ 2024-03-04T19:52:09.422285621Z eval-quality-manual-j98-xbtcz Node.js v18.19.1                                                                                                                                │
│ 2024-03-04T19:52:09.433518837Z eval-quality-manual-j98-xbtcz npm ERR! Lifecycle script `pipeline:conversationQualityCheck` failed with error:                                                                │
│ 2024-03-04T19:52:09.433769583Z eval-quality-manual-j98-xbtcz npm ERR! Error: command failed                                                                                                                  │
│ 2024-03-04T19:52:09.434086722Z eval-quality-manual-j98-xbtcz npm ERR!   in workspace: chatbot-eval-mongodb-public@0.0.0       
```

figure out why..worst case process.exit(0).

that's what we do for the CLI, but probably shoudln't